fish-rust/src/future_feature_flags.rs

```diff
@@ -31,74 +28,49 @@ mod future_feature_flags_ffi {
         ampersand_nobg_in_token,
     }
 
-    /// Metadata about feature flags.
-    struct feature_metadata_t {
-        flag: FeatureFlag,
-        name: UniquePtr<CxxWString>,
-        groups: UniquePtr<CxxWString>,
-        description: UniquePtr<CxxWString>,
-        default_value: bool,
-        read_only: bool,
-    }
-
     extern "Rust" {
-        type Features;
-        fn test(self: &Features, flag: FeatureFlag) -> bool;
-        fn set(self: &mut Features, flag: FeatureFlag, value: bool);
-        fn set_from_string(self: &mut Features, str: wcharz_t);
-        fn fish_features() -> *const Features;
-        fn feature_test(flag: FeatureFlag) -> bool;
-        fn mutable_fish_features() -> *mut Features;
-        fn feature_metadata() -> [feature_metadata_t; 4];
+        #[cxx_name = "feature_test"]
+        fn test(flag: FeatureFlag) -> bool;
+        #[cxx_name = "feature_set"]
+        fn set(flag: FeatureFlag, value: bool);
+        #[cxx_name = "feature_set_from_string"]
+        fn set_from_string(str: wcharz_t);
     }
 }
 
-pub use future_feature_flags_ffi::{feature_metadata_t, FeatureFlag};
+pub use future_feature_flags_ffi::FeatureFlag;
 
-pub struct Features {
+struct Features {
     // Values for the flags.
     // These are atomic to "fix" a race reported by tsan where tests of feature flags and other
     // tests which use them conceptually race.
-    values: [AtomicBool; metadata.len()],
+    values: [bool; METADATA.len()],
 }
 
 /// Metadata about feature flags.
-struct FeatureMetadata {
+pub struct FeatureMetadata {
     /// The flag itself.
-    flag: FeatureFlag,
+    pub flag: FeatureFlag,
```

Comment:
this commit should be split up, it does at least two things (the rwlock change and the refactoring)

Reply:
Done!